// @node-red/runtime/lib/homegear/Homegear.js
var homegear = require("@homegear/homegear-nodejs");
var Log;
var runtime;
var redUtil = require("@node-red/util").util;
var socketPath = '';
var ready = false;
var hg = null;

class Homegear {
    constructor() {
        ready = false;
        hg = new homegear.Homegear(socketPath, this.connected, this.disconnected, null, this.nodeInput)
    }

    get() {
        return hg;
    }

    connected() {
        Log.info("Connected to Homegear.");
        try {
            if (ready) hg.invoke("noderedEvent", ["global", "isReady"]);
        } catch (e) {
            console.log(e);
        }
    }

    disconnected() {
        Log.info("Disconnected from Homegear. Closing process.");
        process.exit(0);
    }

    nodeInput(nodeId, nodeInfo, inputIndex, message, synchronous) {
        try {
            var node = runtime.flows.get(nodeId);
            if (!node) return;

            if (!message.hasOwnProperty('_msgid')) message._msgid = redUtil.generateId();
            if (!message.hasOwnProperty('topic')) message.topic = '';
            node.synchronousOutput = synchronous;
            node.receive(message);
        } catch (e) {
            console.log(e);
        }
    }

    isReady(_ready) {
        try {
            ready = _ready;
            if (_ready && hg.connected()) hg.invoke("noderedEvent", ["global", "isReady"]);
        } catch (e) {
            console.log(e);
        }
    }
}

module.exports = {
    init: function(_runtime) {
        Log = _runtime.log;
        runtime = _runtime;
        socketPath = _runtime.settings.homegearSocketPath || '';
    },
    create: function() {
        return new Homegear();
    },
    Homegear: Homegear
}
